### PR TITLE
fix: use Floor rounding for dragon loss share burn (OSU-1492)

### PR DESCRIPTION
**Pull request by 0xferit**
## Summary

- Changed `Math.Rounding.Ceil` to `Math.Rounding.Floor` in `YieldDonatingTokenizedStrategy._handleDragonLossProtection` (line 93) to prevent over-burning 1 dragon share whenever `loss * totalSupply % totalAssets != 0`
- Updated NatSpec to reflect the corrected rounding direction
- Added regression test `test_burnConversion_floorRoundingAtNonUnityPPS` that operates at PPS < 1, where Floor and Ceil produce different results (all existing burn tests operated at PPS=1 where the difference is invisible)

Addresses Quentin's security audit finding V3 (Medium). Aligns with upstream Yearn V3, sibling `DragonTokenizedStrategy`, and Morpho Blue -- all of which use Floor rounding for loss-to-shares conversion.

### CODE REVIEW

**Comment by 0xferit** [comment]
@codex review

File: test/unit/strategies/yieldDonating/Accounting.t.sol [hunk 1]
```diff
@@ -625,6 +625,81 @@ contract AccountingTest is Setup {
...
+        // 7e18 * 150e18 = 1050e36
+        // 1050e36 / 120e18 = 8.75e18
+        // Floor = 8750000000000000000, Ceil = 8750000000000000001? No.
+        // mulDiv(7e18, 150e18, 120e18, Floor) = (7*150)/120 = 1050/120 = 8.75
+        // In integer math: 7e18 * 150e18 = 1050e36, / 120e18 = 8750000000000000000
+        // That's exact. We need a loss where loss * totalSupply % totalAssets != 0.
+        //
+        // Use loss = 7e18 + 1 (not evenly divisible):
+        // (7e18+1) * 150e18 / 120e18 -- remainder = (7e18+1)*150e18 % 120e18
+        // = 150e18 % 120e18 = 30e18 != 0 -- non-zero remainder
```
[the remark this answers is not in the document]
Comment (0xferit) [reply test/unit/strategies/yieldDonating/Accounting.t.sol]: Agreed -- that was working-out from the derivation that leaked into the final test. Cleaned up in 7be6c678.

File: test/unit/strategies/yieldDonating/Accounting.t.sol [hunk 2]
```diff
@@ -625,6 +625,81 @@ contract AccountingTest is Setup {
...
+        uint256 loss = 7e18 + 1;
+        yieldSource.simulateLoss(loss);
+
+        uint256 dragonSharesBefore = strategy.balanceOf(donationAddress);
+
+        // Compute expected floor shares: loss * totalSupply / totalAssets (integer division = floor)
+        uint256 floorShares = (loss * totalSupply) / totalAssets;
+        uint256 ceilShares = floorShares + 1; // ceil adds 1 when remainder != 0
```
[the remark this answers is not in the document]
Comment (0xferit) [reply test/unit/strategies/yieldDonating/Accounting.t.sol:680]: Intentionally using raw Solidity integer division here. The test needs to independently verify the contract's behavior -- if both the test and the implementation call the same Math.mulDiv, the assertion becomes tautological. The values are small enough (~1050e36) that overflow is not a concern for uint256.

**Comment by 0xferit** [comment]
@codex review

File: src/strategies/yieldDonating/YieldDonatingTokenizedStrategy.sol [hunk 3]
```diff
@@ -90,7 +90,7 @@ contract YieldDonatingTokenizedStrategy is TokenizedStrategy {
     function _handleDragonLossProtection(StrategyData storage S, uint256 loss) internal {
         if (S.enableBurning) {
             // Convert loss to shares that should be burned
-            uint256 sharesToBurn = _convertToShares(S, loss, Math.Rounding.Ceil);
+            uint256 sharesToBurn = _convertToShares(S, loss, Math.Rounding.Floor);
```
Comment (skimaharvey) [inline src/strategies/yieldDonating/YieldDonatingTokenizedStrategy.sol:96]: i would still add a comments that explains why it is fine to round down - as it wont go in favor of the depositors

**Review by skimaharvey — DISMISSED** [review]
Would add a comment that expalins why we can round down also have you looked if we Should we also implement the same rational at YieldSkimming level? cc @qGolem

File: src/strategies/yieldDonating/YieldDonatingTokenizedStrategy.sol (hunk 3 above)
Comment (0xferit) [reply src/strategies/yieldDonating/YieldDonatingTokenizedStrategy.sol:96]: Good call. Added an inline comment in 3b04659c explaining the rationale: any sub-wei remainder from Floor gets socialized via PPS reduction across all holders, rather than over-burning dragon shares which would systematically transfer value from dragon to depositors.

**Review by qGolem — CHANGES_REQUESTED** [review]
Check yield skmming for the same

**Comment by 0xferit** [comment]
> Check yield skmming for the same

Checked. The bug does not exist in YieldSkimmingTokenizedStrategy.

YieldSkimming uses a 1:1 value-debt architecture where 1 share = 1 ETH value. Its loss burn path at [YieldSkimmingTokenizedStrategy.sol:692](https://github.com/golemfoundation/octant-v2-core/blob/e89d05dd/src/strategies/yieldSkimming/YieldSkimmingTokenizedStrategy.sol#L726) is:

  `dragonBurn = Math.min(lossValue, dragonBalance);`

  This is a direct value comparison; no `_convertToShares` call, no division, no rounding. The Ceil-vs-Floor issue only manifests when (loss * totalSupply) % totalAssets != 0 inside _convertToShares, which YieldSkimming never hits during loss handling.

  The bug is specific to YieldDonating's PPS-based share conversion model.

**Review by qGolem — APPROVED** [review]
